hal/lib/athena/handles/IndexedHandleResource.h

```diff
@@ -41,57 +38,40 @@ class IndexedHandleResource {
  public:
   IndexedHandleResource(const IndexedHandleResource&) = delete;
   IndexedHandleResource operator=(const IndexedHandleResource&) = delete;
-  IndexedHandleResource();
-  THandle Allocate(int16_t index, const TStruct& toSet);
-  TStruct Get(THandle handle, int32_t* status);
+  IndexedHandleResource() = default;
+  THandle Allocate(int16_t index);
+  std::shared_ptr<TStruct> Get(THandle handle);
   void Free(THandle handle);
 
  private:
-  TStruct m_structures[size];
-  bool m_allocated[size];
+  std::shared_ptr<TStruct> m_structures[size];
   priority_mutex m_handleMutexes[size];
 };
 
 template <typename THandle, typename TStruct, int16_t size,
           HalHandleEnum enumValue>
-IndexedHandleResource<THandle, TStruct, size,
-                      enumValue>::IndexedHandleResource() {
-  // ensure all allocations are initially false.
-  for (int i = 0; i < size; i++) {
-    m_allocated[i] = false;
-  }
-}
-
-template <typename THandle, typename TStruct, int16_t size,
-          HalHandleEnum enumValue>
 THandle IndexedHandleResource<THandle, TStruct, size, enumValue>::Allocate(
-    int16_t index, const TStruct& toSet) {
+    int16_t index) {
   // don't aquire the lock if we can fail early.
   if (index < 0 || index >= size) return HAL_HANDLE_INDEX_OUT_OF_RANGE;
   std::lock_guard<priority_mutex> sync(m_handleMutexes[index]);
   // check for allocation, otherwise allocate and return a valid handle
-  if (m_allocated[index]) return HAL_HANDLE_ALREADY_ALLOCATED;
-  m_allocated[index] = true;
-  m_structures[index] = toSet;
+  if (m_structures[index] != nullptr) return HAL_HANDLE_ALREADY_ALLOCATED;
+  m_structures[index] = std::make_shared<TStruct>();
   return (THandle)hal::createHandle(index, enumValue);
 }
 
 template <typename THandle, typename TStruct, int16_t size,
           HalHandleEnum enumValue>
-TStruct IndexedHandleResource<THandle, TStruct, size, enumValue>::Get(
-    THandle handle, int32_t* status) {
+std::shared_ptr<TStruct> IndexedHandleResource<THandle, TStruct, size, enumValue>::Get(
+    THandle handle) {
   // get handle index, and fail early if index out of range or wrong handle
   int16_t index = getHandleTypedIndex(handle, enumValue);
   if (index < 0 || index >= size) {
-    *status = IndexedResourceIndexOutOfRange;
-    return TStruct();
+    return nullptr;
```

Comment:
Do we need to know? It would only change the error message, but we would still throw an error anyway. I can add that back in quickly, but I doubt we'll use the difference other then when printing the actual error message.